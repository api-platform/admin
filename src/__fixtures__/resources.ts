import type { Resource } from '@api-platform/api-doc-parser';

const resources: Resource[] = [
  {
    name: 'books',
    url: 'https://demo.api-platform.com/books',
    id: 'http://schema.org/Book',
    title: 'Book',
    fields: [
      {
        name: 'isbn',
        id: 'http://schema.org/isbn',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: false,
        description: 'The ISBN of the book',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'title',
        id: 'http://schema.org/name',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: true,
        description: 'The title of the book',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'description',
        id: 'http://schema.org/description',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: true,
        description: 'A description of the item',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'author',
        id: 'http://schema.org/author',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: true,
        description:
          'The author of this content or rating. Please note that author is special in that HTML 5 provides a special mechanism for indicating authorship via the rel tag. That is equivalent to this and may be used interchangeably',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'publicationDate',
        id: 'http://schema.org/dateCreated',
        range: 'http://www.w3.org/2001/XMLSchema#dateTime',
        reference: null,
        required: true,
        description:
          'The date on which the CreativeWork was created or the item was added to a DataFeed',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'reviews',
        id: 'http://schema.org/reviews',
        range: 'http://schema.org/Review',
        reference: {
          name: 'reviews',
          url: 'https://demo.api-platform.com/reviews',
          id: 'http://schema.org/Review',
          title: 'Review',
          fields: [
            {
              name: 'body',
              id: 'http://schema.org/reviewBody',
              range: 'http://www.w3.org/2001/XMLSchema#string',
              reference: null,
              required: false,
              description: 'The actual body of the review',
              maxCardinality: null,
              deprecated: false,
            },
            {
              name: 'rating',
              id: 'https://demo.api-platform.com/docs.jsonld#Review/rating',
              range: 'http://www.w3.org/2001/XMLSchema#integer',
              reference: null,
              required: false,
              description: 'A rating',
              maxCardinality: null,
              deprecated: false,
            },
            {
              name: 'letter',
              id: 'https://demo.api-platform.com/docs.jsonld#Review/letter',
              range: 'http://www.w3.org/2001/XMLSchema#string',
              reference: null,
              required: false,
              description:
                'DEPRECATED (use rating now): A letter to rate the book',
              maxCardinality: null,
              deprecated: true,
            },
            {
              name: 'book',
              id: 'http://schema.org/itemReviewed',
              range: 'http://schema.org/Book',
              required: true,
              description: 'The item that is being reviewed/rated',
              maxCardinality: 1,
              deprecated: false,
            },
            {
              name: 'author',
              id: 'http://schema.org/author',
              range: 'http://www.w3.org/2001/XMLSchema#string',
              reference: null,
              required: false,
              description: 'Author the author of the review',
              maxCardinality: null,
              deprecated: false,
            },
            {
              name: 'publicationDate',
              id: 'https://demo.api-platform.com/docs.jsonld#Review/publicationDate',
              range: 'http://www.w3.org/2001/XMLSchema#dateTime',
              reference: null,
              required: false,
              description: 'Author the author of the review',
              maxCardinality: null,
              deprecated: false,
            },
          ],
          readableFields: [],
          writableFields: [],
          operations: [
            {
              name: 'Retrieves the collection of Review resources.',
              type: 'list',
              method: 'GET',
              returns: 'http://www.w3.org/ns/hydra/core#Collection',
              types: [
                'http://www.w3.org/ns/hydra/core#Operation',
                'http://schema.org/FindAction',
              ],
              deprecated: false,
            },
            {
              name: 'Creates a Review resource.',
              type: 'create',
              method: 'POST',
              expects: 'http://schema.org/Review',
              returns: 'http://schema.org/Review',
              types: [
                'http://www.w3.org/ns/hydra/core#Operation',
                'http://schema.org/CreateAction',
              ],
              deprecated: false,
            },
            {
              name: 'Retrieves Review resource.',
              type: 'show',
              method: 'GET',
              returns: 'http://schema.org/Review',
              types: [
                'http://www.w3.org/ns/hydra/core#Operation',
                'http://schema.org/FindAction',
              ],
              deprecated: false,
            },
            {
              name: 'Deletes the Review resource.',
              type: 'delete',
              method: 'DELETE',
              returns: 'http://www.w3.org/2002/07/owl#Nothing',
              types: [
                'http://www.w3.org/ns/hydra/core#Operation',
                'http://schema.org/DeleteAction',
              ],
              deprecated: false,
            },
            {
              name: 'Replaces the Review resource.',
              type: 'edit',
              method: 'PUT',
              expects: 'http://schema.org/Review',
              returns: 'http://schema.org/Review',
              types: [
                'http://www.w3.org/ns/hydra/core#Operation',
                'http://schema.org/ReplaceAction',
              ],
              deprecated: false,
            },
            {
              name: 'Updates the Review resource.',
              type: 'edit',
              method: 'PATCH',
              expects: 'http://schema.org/Review',
              returns: 'http://schema.org/Review',
              types: ['http://www.w3.org/ns/hydra/core#Operation'],
              deprecated: false,
            },
          ],
          deprecated: false,
          parameters: [],
        },
        required: false,
        description: "The book's reviews",
        maxCardinality: null,
        deprecated: false,
      },
    ],
    readableFields: [
      {
        name: 'isbn',
        id: 'http://schema.org/isbn',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: false,
        description: 'The ISBN of the book',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'title',
        id: 'http://schema.org/name',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: true,
        description: 'The title of the book',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'description',
        id: 'http://schema.org/description',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: true,
        description: 'A description of the item',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'author',
        id: 'http://schema.org/author',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: true,
        description:
          'The author of this content or rating. Please note that author is special in that HTML 5 provides a special mechanism for indicating authorship via the rel tag. That is equivalent to this and may be used interchangeably',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'publicationDate',
        id: 'http://schema.org/dateCreated',
        range: 'http://www.w3.org/2001/XMLSchema#dateTime',
        reference: null,
        required: true,
        description:
          'The date on which the CreativeWork was created or the item was added to a DataFeed',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'reviews',
        id: 'http://schema.org/reviews',
        range: 'http://schema.org/Review',
        reference: {
          name: 'reviews',
          url: 'https://demo.api-platform.com/reviews',
          id: 'http://schema.org/Review',
          title: 'Review',
          fields: [
            {
              name: 'body',
              id: 'http://schema.org/reviewBody',
              range: 'http://www.w3.org/2001/XMLSchema#string',
              reference: null,
              required: false,
              description: 'The actual body of the review',
              maxCardinality: null,
              deprecated: false,
            },
            {
              name: 'rating',
              id: 'https://demo.api-platform.com/docs.jsonld#Review/rating',
              range: 'http://www.w3.org/2001/XMLSchema#integer',
              reference: null,
              required: false,
              description: 'A rating',
              maxCardinality: null,
              deprecated: false,
            },
            {
              name: 'letter',
              id: 'https://demo.api-platform.com/docs.jsonld#Review/letter',
              range: 'http://www.w3.org/2001/XMLSchema#string',
              reference: null,
              required: false,
              description:
                'DEPRECATED (use rating now): A letter to rate the book',
              maxCardinality: null,
              deprecated: true,
            },
            {
              name: 'book',
              id: 'http://schema.org/itemReviewed',
              range: 'http://schema.org/Book',
              required: true,
              description: 'The item that is being reviewed/rated',
              maxCardinality: 1,
              deprecated: false,
            },
            {
              name: 'author',
              id: 'http://schema.org/author',
              range: 'http://www.w3.org/2001/XMLSchema#string',
              reference: null,
              required: false,
              description: 'Author the author of the review',
              maxCardinality: null,
              deprecated: false,
            },
            {
              name: 'publicationDate',
              id: 'https://demo.api-platform.com/docs.jsonld#Review/publicationDate',
              range: 'http://www.w3.org/2001/XMLSchema#dateTime',
              reference: null,
              required: false,
              description: 'Author the author of the review',
              maxCardinality: null,
              deprecated: false,
            },
          ],
          readableFields: [],
          writableFields: [],
          operations: [
            {
              name: 'Retrieves the collection of Review resources.',
              type: 'list',
              method: 'GET',
              returns: 'http://www.w3.org/ns/hydra/core#Collection',
              types: [
                'http://www.w3.org/ns/hydra/core#Operation',
                'http://schema.org/FindAction',
              ],
              deprecated: false,
            },
            {
              name: 'Creates a Review resource.',
              type: 'create',
              method: 'POST',
              expects: 'http://schema.org/Review',
              returns: 'http://schema.org/Review',
              types: [
                'http://www.w3.org/ns/hydra/core#Operation',
                'http://schema.org/CreateAction',
              ],
              deprecated: false,
            },
            {
              name: 'Retrieves Review resource.',
              type: 'show',
              method: 'GET',
              returns: 'http://schema.org/Review',
              types: [
                'http://www.w3.org/ns/hydra/core#Operation',
                'http://schema.org/FindAction',
              ],
              deprecated: false,
            },
            {
              name: 'Deletes the Review resource.',
              type: 'delete',
              method: 'DELETE',
              returns: 'http://www.w3.org/2002/07/owl#Nothing',
              types: [
                'http://www.w3.org/ns/hydra/core#Operation',
                'http://schema.org/DeleteAction',
              ],
              deprecated: false,
            },
            {
              name: 'Replaces the Review resource.',
              type: 'edit',
              method: 'PUT',
              expects: 'http://schema.org/Review',
              returns: 'http://schema.org/Review',
              types: [
                'http://www.w3.org/ns/hydra/core#Operation',
                'http://schema.org/ReplaceAction',
              ],
              deprecated: false,
            },
            {
              name: 'Updates the Review resource.',
              type: 'edit',
              method: 'PATCH',
              expects: 'http://schema.org/Review',
              returns: 'http://schema.org/Review',
              types: ['http://www.w3.org/ns/hydra/core#Operation'],
              deprecated: false,
            },
          ],
          deprecated: false,
          parameters: [],
        },
        required: false,
        description: "The book's reviews",
        maxCardinality: null,
        deprecated: false,
      },
    ],
    writableFields: [
      {
        name: 'isbn',
        id: 'http://schema.org/isbn',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: false,
        description: 'The ISBN of the book',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'title',
        id: 'http://schema.org/name',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: true,
        description: 'The title of the book',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'description',
        id: 'http://schema.org/description',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: true,
        description: 'A description of the item',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'author',
        id: 'http://schema.org/author',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: true,
        description:
          'The author of this content or rating. Please note that author is special in that HTML 5 provides a special mechanism for indicating authorship via the rel tag. That is equivalent to this and may be used interchangeably',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'publicationDate',
        id: 'http://schema.org/dateCreated',
        range: 'http://www.w3.org/2001/XMLSchema#dateTime',
        reference: null,
        required: true,
        description:
          'The date on which the CreativeWork was created or the item was added to a DataFeed',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'reviews',
        id: 'http://schema.org/reviews',
        range: 'http://schema.org/Review',
        reference: {
          name: 'reviews',
          url: 'https://demo.api-platform.com/reviews',
          id: 'http://schema.org/Review',
          title: 'Review',
          fields: [
            {
              name: 'body',
              id: 'http://schema.org/reviewBody',
              range: 'http://www.w3.org/2001/XMLSchema#string',
              reference: null,
              required: false,
              description: 'The actual body of the review',
              maxCardinality: null,
              deprecated: false,
            },
            {
              name: 'rating',
              id: 'https://demo.api-platform.com/docs.jsonld#Review/rating',
              range: 'http://www.w3.org/2001/XMLSchema#integer',
              reference: null,
              required: false,
              description: 'A rating',
              maxCardinality: null,
              deprecated: false,
            },
            {
              name: 'letter',
              id: 'https://demo.api-platform.com/docs.jsonld#Review/letter',
              range: 'http://www.w3.org/2001/XMLSchema#string',
              reference: null,
              required: false,
              description:
                'DEPRECATED (use rating now): A letter to rate the book',
              maxCardinality: null,
              deprecated: true,
            },
            {
              name: 'book',
              id: 'http://schema.org/itemReviewed',
              range: 'http://schema.org/Book',
              required: true,
              description: 'The item that is being reviewed/rated',
              maxCardinality: 1,
              deprecated: false,
            },
            {
              name: 'author',
              id: 'http://schema.org/author',
              range: 'http://www.w3.org/2001/XMLSchema#string',
              reference: null,
              required: false,
              description: 'Author the author of the review',
              maxCardinality: null,
              deprecated: false,
            },
            {
              name: 'publicationDate',
              id: 'https://demo.api-platform.com/docs.jsonld#Review/publicationDate',
              range: 'http://www.w3.org/2001/XMLSchema#dateTime',
              reference: null,
              required: false,
              description: 'Author the author of the review',
              maxCardinality: null,
              deprecated: false,
            },
          ],
          readableFields: [],
          writableFields: [],
          operations: [
            {
              name: 'Retrieves the collection of Review resources.',
              type: 'list',
              method: 'GET',
              returns: 'http://www.w3.org/ns/hydra/core#Collection',
              types: [
                'http://www.w3.org/ns/hydra/core#Operation',
                'http://schema.org/FindAction',
              ],
              deprecated: false,
            },
            {
              name: 'Creates a Review resource.',
              type: 'create',
              method: 'POST',
              expects: 'http://schema.org/Review',
              returns: 'http://schema.org/Review',
              types: [
                'http://www.w3.org/ns/hydra/core#Operation',
                'http://schema.org/CreateAction',
              ],
              deprecated: false,
            },
            {
              name: 'Retrieves Review resource.',
              type: 'show',
              method: 'GET',
              returns: 'http://schema.org/Review',
              types: [
                'http://www.w3.org/ns/hydra/core#Operation',
                'http://schema.org/FindAction',
              ],
              deprecated: false,
            },
            {
              name: 'Deletes the Review resource.',
              type: 'delete',
              method: 'DELETE',
              returns: 'http://www.w3.org/2002/07/owl#Nothing',
              types: [
                'http://www.w3.org/ns/hydra/core#Operation',
                'http://schema.org/DeleteAction',
              ],
              deprecated: false,
            },
            {
              name: 'Replaces the Review resource.',
              type: 'edit',
              method: 'PUT',
              expects: 'http://schema.org/Review',
              returns: 'http://schema.org/Review',
              types: [
                'http://www.w3.org/ns/hydra/core#Operation',
                'http://schema.org/ReplaceAction',
              ],
              deprecated: false,
            },
            {
              name: 'Updates the Review resource.',
              type: 'edit',
              method: 'PATCH',
              expects: 'http://schema.org/Review',
              returns: 'http://schema.org/Review',
              types: ['http://www.w3.org/ns/hydra/core#Operation'],
              deprecated: false,
            },
          ],
          deprecated: false,
          parameters: [],
        },
        required: false,
        description: "The book's reviews",
        maxCardinality: null,
        deprecated: false,
      },
    ],
    operations: [
      {
        name: 'Retrieves the collection of Book resources.',
        type: 'list',
        method: 'GET',
        returns: 'http://www.w3.org/ns/hydra/core#Collection',
        types: [
          'http://www.w3.org/ns/hydra/core#Operation',
          'http://schema.org/FindAction',
        ],
        deprecated: false,
      },
      {
        name: 'Creates a Book resource.',
        type: 'create',
        method: 'POST',
        expects: 'http://schema.org/Book',
        returns: 'http://schema.org/Book',
        types: [
          'http://www.w3.org/ns/hydra/core#Operation',
          'http://schema.org/CreateAction',
        ],
        deprecated: false,
      },
      {
        name: 'Retrieves Book resource.',
        type: 'show',
        method: 'GET',
        returns: 'http://schema.org/Book',
        types: [
          'http://www.w3.org/ns/hydra/core#Operation',
          'http://schema.org/FindAction',
        ],
        deprecated: false,
      },
      {
        name: 'Replaces the Book resource.',
        type: 'edit',
        method: 'PUT',
        expects: 'http://schema.org/Book',
        returns: 'http://schema.org/Book',
        types: [
          'http://www.w3.org/ns/hydra/core#Operation',
          'http://schema.org/ReplaceAction',
        ],
        deprecated: false,
      },
      {
        name: 'Deletes the Book resource.',
        type: 'delete',
        method: 'DELETE',
        returns: 'http://www.w3.org/2002/07/owl#Nothing',
        types: [
          'http://www.w3.org/ns/hydra/core#Operation',
          'http://schema.org/DeleteAction',
        ],
        deprecated: false,
      },
    ],
    deprecated: false,
    parameters: [],
  },
  {
    name: 'parchments',
    url: 'https://demo.api-platform.com/parchments',
    id: 'https://demo.api-platform.com/docs.jsonld#Parchment',
    title: 'Parchment',
    fields: [
      {
        name: 'title',
        id: 'https://demo.api-platform.com/docs.jsonld#Parchment/title',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: true,
        description: 'The title of the book',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'description',
        id: 'https://demo.api-platform.com/docs.jsonld#Parchment/description',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: true,
        description: 'A description of the item',
        maxCardinality: null,
        deprecated: false,
      },
    ],
    readableFields: [
      {
        name: 'title',
        id: 'https://demo.api-platform.com/docs.jsonld#Parchment/title',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: true,
        description: 'The title of the book',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'description',
        id: 'https://demo.api-platform.com/docs.jsonld#Parchment/description',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: true,
        description: 'A description of the item',
        maxCardinality: null,
        deprecated: false,
      },
    ],
    writableFields: [
      {
        name: 'title',
        id: 'https://demo.api-platform.com/docs.jsonld#Parchment/title',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: true,
        description: 'The title of the book',
        maxCardinality: null,
        deprecated: false,
      },
      {
        name: 'description',
        id: 'https://demo.api-platform.com/docs.jsonld#Parchment/description',
        range: 'http://www.w3.org/2001/XMLSchema#string',
        reference: null,
        required: true,
        description: 'A description of the item',
        maxCardinality: null,
        deprecated: false,
      },
    ],
    operations: [
      {
        name: 'Retrieves the collection of Parchment resources.',
        type: 'list',
        method: 'GET',
        returns: 'http://www.w3.org/ns/hydra/core#Collection',
        types: [
          'http://www.w3.org/ns/hydra/core#Operation',
          'http://schema.org/FindAction',
        ],
        deprecated: true,
      },
      {
        name: 'Creates a Parchment resource.',
        type: 'create',
        method: 'POST',
        expects: 'https://demo.api-platform.com/docs.jsonld#Parchment',
        returns: 'https://demo.api-platform.com/docs.jsonld#Parchment',
        types: [
          'http://www.w3.org/ns/hydra/core#Operation',
          'http://schema.org/CreateAction',
        ],
        deprecated: true,
      },
      {
        name: 'Retrieves Parchment resource.',
        type: 'show',
        method: 'GET',
        returns: 'https://demo.api-platform.com/docs.jsonld#Parchment',
        types: [
          'http://www.w3.org/ns/hydra/core#Operation',
          'http://schema.org/FindAction',
        ],
        deprecated: true,
      },
      {
        name: 'Deletes the Parchment resource.',
        type: 'delete',
        method: 'DELETE',
        returns: 'http://www.w3.org/2002/07/owl#Nothing',
        types: [
          'http://www.w3.org/ns/hydra/core#Operation',
          'http://schema.org/DeleteAction',
        ],
        deprecated: true,
      },
      {
        name: 'Replaces the Parchment resource.',
        type: 'edit',
        method: 'PUT',
        expects: 'https://demo.api-platform.com/docs.jsonld#Parchment',
        returns: 'https://demo.api-platform.com/docs.jsonld#Parchment',
        types: [
          'http://www.w3.org/ns/hydra/core#Operation',
          'http://schema.org/ReplaceAction',
        ],
        deprecated: true,
      },
      {
        name: 'Updates the Parchment resource.',
        type: 'edit',
        method: 'PATCH',
        expects: 'https://demo.api-platform.com/docs.jsonld#Parchment',
        returns: 'https://demo.api-platform.com/docs.jsonld#Parchment',
        types: ['http://www.w3.org/ns/hydra/core#Operation'],
        deprecated: true,
      },
    ],
    deprecated: true,
    parameters: [],
  },
];

export default resources;
