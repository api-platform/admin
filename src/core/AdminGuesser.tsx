import React, { useEffect, useMemo, useState } from 'react';
import {
  AdminContext,
  defaultI18nProvider,
  /* tree-shaking no-side-effects-when-called */ localStorageStore,
} from 'react-admin';

import type { ComponentType } from 'react';
import type { AdminProps } from 'react-admin';
import type { Resource } from '@api-platform/api-doc-parser';

import { AdminResourcesGuesser } from './AdminResourcesGuesser.js';
import IntrospectionContext from '../introspection/IntrospectionContext.js';
import SchemaAnalyzerContext from '../introspection/SchemaAnalyzerContext.js';
import {
  Error as DefaultError,
  Layout,
  LoginPage,
  darkTheme as defaultDarkTheme,
  lightTheme as defaultLightTheme,
} from '../layout/index.js';
import type { ApiPlatformAdminDataProvider, SchemaAnalyzer } from '../types.js';

export interface AdminGuesserProps extends AdminProps {
  admin?: ComponentType<AdminProps>;
  dataProvider: ApiPlatformAdminDataProvider;
  schemaAnalyzer: SchemaAnalyzer;
  includeDeprecated?: boolean;
}

const defaultStore = localStorageStore();

const AdminGuesser = ({
  // Props for SchemaAnalyzerContext
  schemaAnalyzer,
  // Props for AdminResourcesGuesser
  includeDeprecated = false,
  // Admin props
  basename,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  error = DefaultError as any,
  store = defaultStore,
  dataProvider,
  i18nProvider = defaultI18nProvider,
  authProvider,
  queryClient,
  defaultTheme,
  layout = Layout,
  loginPage = LoginPage,
  loading: loadingPage,
  theme = defaultLightTheme,
  darkTheme = defaultDarkTheme,
  // Other props
  children,
  ...rest
}: AdminGuesserProps) => {
  const [resources, setResources] = useState<Resource[]>([]);
  const [loading, setLoading] = useState(true);
  const [, setError] = useState();
  const [introspect, setIntrospect] = useState(true);

  useEffect(() => {
    if (typeof dataProvider.introspect !== 'function') {
      throw new Error(
        'The given dataProvider needs to expose an "introspect" function returning a parsed API documentation from api-doc-parser',
      );
    }

    if (!introspect) {
      return;
    }

    dataProvider
      .introspect()
      .then(({ data }) => {
        setResources(data.resources ?? []);
        setIntrospect(false);
        setLoading(false);
      })
      .catch((err) => {
        // Allow err to be caught by the error boundary
        setError(() => {
          throw err;
        });
      });
  }, [introspect, dataProvider]);

  const introspectionContext = useMemo(
    () => ({
      introspect: () => {
        setLoading(true);
        setIntrospect(true);
      },
    }),
    [setLoading, setIntrospect],
  );

  return (
    <AdminContext
      i18nProvider={i18nProvider}
      dataProvider={dataProvider}
      basename={basename}
      authProvider={authProvider}
      store={store}
      queryClient={queryClient}
      theme={theme}
      darkTheme={darkTheme}
      defaultTheme={defaultTheme}>
      <IntrospectionContext.Provider value={introspectionContext}>
        <SchemaAnalyzerContext.Provider value={schemaAnalyzer}>
          <AdminResourcesGuesser
            includeDeprecated={includeDeprecated}
            resources={resources}
            loading={loading}
            dataProvider={dataProvider}
            layout={layout}
            loginPage={loginPage}
            loadingPage={loadingPage}
            theme={theme}
            error={error}
            {...rest}>
            {children}
          </AdminResourcesGuesser>
        </SchemaAnalyzerContext.Provider>
      </IntrospectionContext.Provider>
    </AdminContext>
  );
};

export default AdminGuesser;
