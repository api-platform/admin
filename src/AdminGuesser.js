import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import {
  AdminContext,
  AdminUI,
  ComponentPropType,
  Error as DefaultError,
  Loading,
  TranslationProvider,
  defaultI18nProvider,
} from 'react-admin';
import { createHashHistory } from 'history';
import { createMuiTheme } from '@material-ui/core';

import ErrorBoundary from './ErrorBoundary';
import ResourceGuesser from './ResourceGuesser';
import SchemaAnalyzerContext from './SchemaAnalyzerContext';
import { Layout } from './layout';
import introspectReducer from './introspectReducer';

const displayOverrideCode = (resources) => {
  if (process.env.NODE_ENV === 'production') return;

  let code =
    'If you want to override at least one resource, paste this content in the <AdminGuesser> component of your app:\n\n';

  resources.forEach((r) => {
    code += `<ResourceGuesser name={"${r.name}"} />\n`;
  });
  console.info(code);
};

/**
 * AdminResourcesGuesser automatically renders an `<AdminUI>` component for resources exposed by a web API documented with Hydra, OpenAPI or any other format supported by `@api-platform/api-doc-parser`.
 * If child components are passed (usually `<ResourceGuesser>` or `<Resource>` components, but it can be any other React component), they are rendered in the given order.
 * If no children are passed, a `<ResourceGuesser>` component is created for each resource type exposed by the API, in the order they are specified in the API documentation.
 */
export const AdminResourcesGuesser = ({
  children,
  includeDeprecated,
  resources,
  loading,
  loadingPage: LoadingPage = Loading,
  ...rest
}) => {
  if (loading) {
    return <LoadingPage />;
  }

  let resourceChildren = children;
  if (!resourceChildren && resources) {
    const guessResources = includeDeprecated
      ? resources
      : resources.filter((r) => !r.deprecated);
    resourceChildren = guessResources.map((r) => (
      <ResourceGuesser name={r.name} key={r.name} />
    ));
    displayOverrideCode(guessResources);
  }

  return (
    <AdminUI loading={LoadingPage} {...rest}>
      {resourceChildren}
    </AdminUI>
  );
};

const defaultTheme = createMuiTheme({
  palette: {
    primary: {
      contrastText: '#ffffff',
      main: '#38a9b4',
    },
    secondary: {
      main: '#288690',
    },
  },
});

const AdminGuesser = ({
  // Props for SchemaAnalyzerContext
  schemaAnalyzer,
  // Props for AdminContext
  dataProvider,
  authProvider,
  i18nProvider,
  history,
  customReducers = {},
  customSagas,
  initialState,
  // Props for AdminResourcesGuesser
  includeDeprecated = false,
  // Props for AdminUI
  customRoutes = [],
  appLayout,
  layout = Layout,
  loginPage,
  loading: loadingPage,
  locale,
  theme = defaultTheme,
  // Other props
  children,
  ...rest
}) => {
  const [resources, setResources] = useState();
  const [loading, setLoading] = useState(true);
  const [, setError] = useState();
  const [addedCustomRoutes, setAddedCustomRoutes] = useState([]);

  if (!history) {
    history = typeof window === 'undefined' ? {} : createHashHistory();
  }

  if (appLayout && process.env.NODE_ENV !== 'production') {
    console.warn(
      'You are using deprecated prop "appLayout", it was replaced by "layout", see https://github.com/marmelab/react-admin/issues/2918',
    );
  }
  if (loginPage === true && process.env.NODE_ENV !== 'production') {
    console.warn(
      'You passed true to the loginPage prop. You must either pass false to disable it or a component class to customize it',
    );
  }
  if (locale && process.env.NODE_ENV !== 'production') {
    console.warn(
      'You are using deprecated prop "locale". You must now pass the initial locale to your i18nProvider',
    );
  }

  useEffect(() => {
    if (typeof dataProvider.introspect !== 'function') {
      throw new Error(
        'The given dataProvider needs to expose an "introspect" function returning a parsed API documentation from api-doc-parser',
      );
    }

    dataProvider
      .introspect()
      .then(({ data, customRoutes }) => {
        setResources(data.resources);
        setAddedCustomRoutes(customRoutes);
        setLoading(false);
      })
      .catch((error) => {
        // Allow error to be caught by the error boundary
        setError(() => {
          throw error;
        });
      });
  }, []);

  return (
    <SchemaAnalyzerContext.Provider value={schemaAnalyzer}>
      <AdminContext
        authProvider={authProvider}
        dataProvider={dataProvider}
        i18nProvider={i18nProvider}
        history={history}
        customReducers={{ introspect: introspectReducer, ...customReducers }}
        customSagas={customSagas}
        initialState={initialState}>
        <AdminResourcesGuesser
          includeDeprecated={includeDeprecated}
          resources={resources}
          customRoutes={[...addedCustomRoutes, ...customRoutes]}
          loading={loading}
          layout={appLayout || layout}
          loginPage={loginPage}
          loadingPage={loadingPage}
          theme={theme}
          {...rest}>
          {children}
        </AdminResourcesGuesser>
      </AdminContext>
    </SchemaAnalyzerContext.Provider>
  );
};

AdminGuesser.propTypes = {
  dataProvider: PropTypes.oneOfType([PropTypes.object, PropTypes.func])
    .isRequired,
  authProvider: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
  i18nProvider: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
  history: PropTypes.object,
  customReducers: PropTypes.object,
  customSagas: PropTypes.object,
  initialState: PropTypes.object,
  schemaAnalyzer: PropTypes.object.isRequired,
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  theme: PropTypes.object,
  includeDeprecated: PropTypes.bool,
  customRoutes: PropTypes.array,
};

const AdminGuesserWithError = ({ error, ...props }) => (
  <TranslationProvider i18nProvider={props.i18nProvider}>
    <ErrorBoundary error={error}>
      <AdminGuesser {...props} />
    </ErrorBoundary>
  </TranslationProvider>
);

AdminGuesserWithError.defaultProps = {
  error: DefaultError,
  i18nProvider: defaultI18nProvider,
};

AdminGuesserWithError.propTypes = {
  error: ComponentPropType,
};

export default AdminGuesserWithError;
