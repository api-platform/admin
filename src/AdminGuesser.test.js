import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { AdminResourcesGuesser } from './AdminGuesser';
import ResourceGuesser from './ResourceGuesser';
import resources from './__fixtures__/resources';

describe('<AdminGuesser />', () => {
  const renderer = new ShallowRenderer();

  test('renders loading', () => {
    const tree = renderer.render(<AdminResourcesGuesser loading={true} />);

    expect(tree).toMatchSnapshot();
  });

  test('renders without custom resources', () => {
    const tree = renderer.render(
      <AdminResourcesGuesser resources={resources} loading={false} />,
    );

    expect(tree).toMatchSnapshot();
  });

  test('renders with custom resources', () => {
    const tree = renderer.render(
      <AdminResourcesGuesser resources={resources} loading={false}>
        <ResourceGuesser name="custom" />
      </AdminResourcesGuesser>,
    );

    expect(tree).toMatchSnapshot();
  });

  test('renders without data', () => {
    const tree = renderer.render(<AdminResourcesGuesser loading={false} />);

    expect(tree).toMatchSnapshot();
  });
});
