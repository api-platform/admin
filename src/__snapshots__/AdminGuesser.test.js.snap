// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminGuesser /> renders errors 1`] = `
<div>
  Error while reading the API schema
</div>
`;

exports[`<AdminGuesser /> renders loading 1`] = `
<Connect(TranslationProviderView)>
  <WithStyles(translate(Loading)) />
</Connect(TranslationProviderView)>
`;

exports[`<AdminGuesser /> renders without custom resources 1`] = `
<withContext(CoreAdminBase)
  appLayout={[Function]}
  catchAll={[Function]}
  loading={[Function]}
  loginPage={[Function]}
  logoutButton={[Function]}
>
  <ResourceGuesser
    name="books"
  />
</withContext(CoreAdminBase)>
`;
